### ISSUE TITLE
Inference on a batch of sequences of variable lengths has stopped working

### ISSUE THREAD

**Issue by xgirones** [microsoft/onnxruntime#1360]
**Describe the bug**
I have a very simple model that takes a tensor containing a padded list of sequences of variable length with shape (minibatch_size, 1, max_num_rows, 24) plus a vector of lengths with shape (minibatch_size), and returns a tensor of predictions with shape  (minibatch_size, max_num_rows, 39). 

The model was trained with Pytorch, and later exported to ONNX using the Pytorch ONNX exporter. It worked fine with the ORT shared library compiled from source a month ago. However, I recompiled the library from code pulled from the ORT repository 3 days ago (Commit ID: 2714576d0ae89f553874394d12c1f3cb4724a6de) and it has stopped working. Somehow it appears that ORT is not able to infer the correct input shape from the supplied input tensor, and it is now using the dummy values from the ONNX export.

```
OrtRun failed . Error description: e:\repo\ort\onnxruntime\core\providers\cpu\tensor\reshape_helper.h:38 onnxruntime::ReshapeHelper::ReshapeHelper (input_shape.Size() % size) == 0 was false. The input tensor cannot be reshaped to the requested shape. Input shape:{16,14,6,24}, requested shape:{-1,5,144}
Stacktrace:
e:\repo\ort\onnxruntime\core\providers\cpu\tensor\reshape.h(34): onnxruntime::Reshape::Compute
e:\repo\ort\onnxruntime\core\framework\sequential_executor.cc(119): onnxruntime::SequentialExecutor::Execute
e:\repo\ort\onnxruntime\core\framework\utils.cc(514): onnxruntime::utils::ExecuteGraph
e:\repo\ort\onnxruntime\core\session\inference_session.cc(707): onnxruntime::InferenceSession::Run
e:\repo\ort\onnxruntime\core\session\onnxruntime_c_api.cc(461): OrtRun
0x140428993 (Unknown symbol)
0x1403ac779 (Unknown symbol)
0x1403b5a1e (Unknown symbol)
0x1403e9821 (Unknown symbol)
0x140506444 (Unknown symbol)
0x1405062ee (Unknown symbol)
0x1405061ae (Unknown symbol)
0x1405064d9 (Unknown symbol)
???: BaseThreadInitThunk
???: RtlUserThreadStart
```

The code I use to initialize the model is

```
  # Model constructor
    def __init__(self, numCols, numClasses, enableSoftmax = False):
        
        # Call the base constructor
        super().__init__()
        
        # Input image dimensions
        self.cnn_downscale_factor = 4
        self.num_cols = numCols
        self.num_classes = numClasses
        self.softmax_enabled = enableSoftmax
        
        # Initialize convolutional blocks.
        self.conv1 = nn.Conv2d( in_channels=1, out_channels=12, kernel_size=3, stride=2, padding=1 )
        self.conv2 = nn.Conv2d( in_channels=self.conv1.out_channels, out_channels=24, kernel_size=3, stride=2, padding=1 )
        
        # Initialize RNN
        rnn_dim = 128
        dense_dim = 96       
        
        self.rnn_input_size = self.conv2.out_channels * self.num_cols // self.cnn_downscale_factor
        #self.rnn = nn.GRU( input_size=self.rnn_input_size, hidden_size=rnn_dim//2, num_layers=1, batch_first=True, bidirectional=True )
        self.rnn = nn.LSTM( input_size=self.rnn_input_size, hidden_size=rnn_dim//2, num_layers=1, batch_first=True, bidirectional=True )
        
        # Initialize dense stages
        self.dense1 = nn.Linear( rnn_dim, dense_dim )
        self.dense2 = nn.Linear( dense_dim, self.num_classes * self.cnn_downscale_factor )
```
An this is the code of the forward pass:

```
# Forward pass
    def forward(self, x, lengths):
        
        # Convolutional stage
        h = self.conv1( x )
        h = my_conv_activation( h )
        h = self.conv2( h )
        h = my_conv_activation( h )
        h = h.permute(0,2,3,1)
        h = h.reshape( -1, h.shape[1], self.rnn_input_size )
        
        # RNN stage
       
        # pack_padded_sequence so that padded items in the sequence won't be shown to the LSTM
        # the lengths tensor must be already divided by the granularity of the CNN stage (4)
        h = torch.nn.utils.rnn.pack_padded_sequence( h, lengths, batch_first=True, enforce_sorted=True )
        
        # run through RNN
        h, _ = self.rnn( h ) 
        
        # undo the packing operation
        h, _ = torch.nn.utils.rnn.pad_packed_sequence( h, batch_first=True )
        
        # Dense stages
        h = self.dense1( h )
        h = my_dense_activation( h )
        
        h = self.dense2( h )
        h = h.reshape( -1, x.shape[2], self.num_classes )
        
        # Softmax for inference, log-softmax for trainning
        if self.softmax_enabled:
            h = F.softmax( h, dim=-1 )
        else:
            h = F.log_softmax( h, dim=-1 )
        
        return h
```
I export the model to ONNX using the following lines

```
dummy_inputs, _, dummy_lengths = generate_padded_minibatch_for_pytorch( sampler, device )
 torch.onnx.export( self, (dummy_inputs, dummy_lengths), pathModel, verbose=False, input_names = [ "padded_sequences", "lengths" ], output_names = [ "output" ])
```

And this is the exported ONNX file

[batch_of_sequences_issue.zip](https://github.com/microsoft/onnxruntime/files/3370308/batch_of_sequences_issue.zip)

Can you please take a look at it? It worked great with the ORT code from a few weeks ago (unfortunately I do not remember the commit as I unwittingly deleted the branch). 

**Urgency**
I am stuck with this.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: Compiled from source 
- Commit ID: 2714576d0ae89f553874394d12c1f3cb4724a6de
- Python version: 3.6
- Visual Studio version (if applicable): 2017
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: Issue also occurs with CPU
- GPU model and memory:

**To Reproduce**
Describe steps/code to reproduce the behavior: Supply a batch of variable length sequences to the model.

**Expected behavior**
I expect ORT to be able to extract the correct tensor shapes from the input and run successfully as it did when using the code pulled a few weeks ago.

**Screenshots**

**Additional context**
[batch_of_sequences_issue.zip](https://github.com/microsoft/onnxruntime/files/3370308/batch_of_sequences_issue.zip)

**Comment by skottmckay** [comment]
I created some random input for padded_sequences with shape 1, 1, 20, 24 (matching the shape defined in the model) and a single entry in 'lengths' with value of '20' which I expected should be <= the third dimension in padded_sequences. When it got to the LSTM node the sequence length from the first input is 5, so the value of 20 in the 'lengths' array was incorrect. What is the expected sequence length at that point?

Can you provide the shape of the padded_sequences value you were using and the actual 'lengths' values?

I would have expected the model inputs to use a dim_param for the unknown dimensions so a concrete shape of {1, 1, 20, 24} for padded_sequences is not what I would have expected. Something like { -1, 1, -1, 24 } where the -1 values are dim_param entries would be more normal in this case. Possibly there's an issue with the exporter causing that.

**Comment by xgirones** [comment]
Thanks for your response. 

Sorry, I forgot to mention that the values in the 'lengths' vector should contain the lengths of the sequences after being processed by the two CNN layers, just before being input to the LSTM. Hence, the original values should be divided by 4 (the lengths of the original sequences must be multiple of 4), which is the CNN granularity . In your experiment, the value in 'lengths' should be 5 (20/4). 

I have tried the same model on branches created from different commits. The code corresponding to the commit a92998c235ee280013ac6ee0fd304a7458b4d0c8

```
Commit a92998c2
by Dmitri Smirnov, 06/10/2019 11:36 PM
committed by GitHub
parent f72c1501
Uncomment ConstantOfShape tests. (#1059)

```
already manifests the issue. On the other hand, the code from commit 7cd2d9f3c41ede8427fb5c4270eceae678e8638a works as expected

```
Commit 7cd2d9f3
by Ryan Hill, 06/05/2019 08:05 PM
committed by GitHub
parent 7c4494a0
child c6abb17b

Change Path_lib to not use Shlwapi.dll if compiled as WINAPI_PARTITION_APP (#1161)
```
Aside from the issue with reshape, I think that there are more problems in the model and/or ORT. In order to get a decent performance I have to compile the library without MKLML

`build.bat --config RelWithDebInfo --use_mkldnn --use_cuda --build_shared_lib  --cudnn_home %CUDNN_PATH% --cuda_home "%CUDA_PATH%"`

When I compile with MKLML

`build.bat --config RelWithDebInfo --use_mkldnn --use_mklml --use_cuda --build_shared_lib  --cudnn_home %CUDNN_PATH% --cuda_home "%CUDA_PATH%"`

it takes minutes for a single inference, and the execution is mostly concentrated in a spin lock (BTW, wouldn't a condition variable be more appropriate in this case as the operations are not trivial?)

```
rnn_helpers.h, line 241

int totalTasks = max / (step > 0 ? step : 1) + (max % step > 0 ? 1 : 0);
while (done != totalTasks) ;
```
In addition, when I run the model on GPU the accuracy is much worse than with CPU. I am compiling the library with support for CUDA 10.1, and this problem does not occur when I use the same library on other models.

**Comment by xgirones** [comment]
> I would have expected the model inputs to use a dim_param for the unknown dimensions so a concrete shape of {1, 1, 20, 24} for padded_sequences is not what I would have expected. Something like { -1, 1, -1, 24 } where the -1 values are dim_param entries would be more normal in this case. Possibly there's an issue with the exporter causing that.

I can try this. How can I make the PyTorch ONNX exporter label a dimension as unknown?

**Comment by hariharans29** [comment]
Tagging @spandantiwari @lara-hdr for PyTorch exporter expertise :)

**Comment by xgirones** [comment]
I have more information: It appears that the problem started with commit 32c6c71e8678df299f06fbeeef26f46646a830bf.

```
Commit 32c6c71e
by Konstantinos Karanasos, 06/07/2019 11:15 PM
committed by Changming Sun
parent cdb27de0
child ccab8165

Convert Shape operator to initializer (#1159)

This PR introduces a rewrite rule that replaces a Shape node with an initializer when the shape of the input is statically known through shape inference.
```
Any idea?

**Comment by skottmckay** [comment]
The root cause will be the pytorch conversion using a concrete shape instead of having symbolic dimensions. That results in the ORT optimizer believing these values will be fixed at runtime, and it replaces nodes where input to a node is constant. 

If the shapes in the model had one or more symbolic dimensions the optimizer would not touch them. 

Workaround pending some insight from the pytorch converter folks would be to disable optimizations by providing a SessionOptions instance to the InferenceSession constructor with graph_optimization_level set to 0.

>>> import onnxruntime as rt
>>> so = rt.SessionOptions()
>>> so.set_graph_optimization_level(0)
>>> so.graph_optimization_level
0
>>> sess = rt.InferenceSession('model.onnx', so)

**Comment by xgirones** [comment]
Would be possible to edit the ONNX model and replace the incorrect fixed dimensions by symbolic tags?

**Comment by xgirones** [comment]
I managed to replace the incorrect constant dimensions by 'dim_param' entries in the ONNX model using the following code in Python:

```
import onnx
from onnx import optimizer

def find_node( model_, type_, name_ ) :
    l = list(filter( lambda x: x.name == name_, getattr(model_.graph,type_)))
    assert( len(l) == 1 )
    return l[ 0 ]

def fix_model( path_src_, path_dst_ ) :
    
    #Load model from disk
    model = onnx.load(path_src_)
    
    # Fix padded sequences
    padded_sequences = find_node(model,"input","padded_sequences")
    shape = padded_sequences.type.tensor_type.shape.dim
    shape[ 0 ].dim_param = "dim_param_0"  # Batch axis
    shape[ 2 ].dim_param = "dim_param_1"  # Sequence axis
    
    # Fix lengths
    lengths = find_node(model,"input","lengths")
    shape = lengths.type.tensor_type.shape.dim
    shape[ 0 ].dim_param = "dim_param_2"  # Batch axis
    
    # Fix output
    output = find_node(model,"output","output")
    shape = output.type.tensor_type.shape.dim
    shape[ 0 ].dim_param = "dim_param_3"  # Batch axis
    shape[ 1 ].dim_param = "dim_param_4"  # Sequence axis
    
    # Tidy up
    onnx.checker.check_model(model)
    model = onnx.optimizer.optimize( model )
    
    #Save the corrected model
    onnx.save( model, path_dst_ )
```

After running the 'fix_model' function the resulting model is: [batch_of_sequences_dim_param.zip](https://github.com/microsoft/onnxruntime/files/3383394/batch_of_sequences_dim_param.zip)

 Now I am able to run the model on the code pulled today from the master branch. However, I think there is something wrong in the GPU implementation. When supplied with a tensor containing sequences of varying lengths the output of the GPU provider is very different from the one obtained with CPU. The divergence corresponding to the longest sequence is the batch is small, though.

 This occurs in all branches I have tested, including the current master branch. I am sure that the GPU implementation is the one that has the problem because when compared to the ground truth, the error of the CPU provider is of the same order as the one I got from PyTorch during training. In addition, when using batches containing a single sequence the results obtained from both GPU and CPU providers are comparable.

Another observation I would like to make is that when evaluating the model on GPU using the code in master, the execution takes a 50% longer than when employing the code pulled from commit 7cd2d9f3c41ede8427fb5c4270eceae678e8638a, for instance.

**Comment by linkerzhang** [comment]
re: "Would it be possible to edit the ONNX model and replace the incorrect constant dimensions by symbolic tags?"
there is no such tool right now :).

re: " However, I think there is something wrong in the GPU implementation. When supplied with a tensor containing sequences of varying lengths the output of the GPU provider is very different from the one obtained with CPU. The divergence corresponding to the longest sequence is the batch is small, though." this is the same issue as you issued #1400, right? 

re: "Another observation I would like to make is that when evaluating the model on GPU using the code in master, the execution takes a 50% longer than when employing the code pulled from commit 7cd2d9f, for instance." thanks for heads up. we'll check it in detail.

**Comment by xgirones** [comment]
> there is no such tool right now :).
I tried editing the model using the ONNX API. So far the ORT does not complain, but it would be great if the PyTorch ONNX exporter could provide a way to specify which axes are dynamic.

> this is the same issue as you issued #1400, right?
Yes, I opened a new issue because it did not appear to be related to the PyTorch exporter.

> thanks for heads up. we'll check it in detail.
Thanks, I will keep posting any finding that might help to improve the product.
